### ISSUE TITLE
ninja backend + Visual Studio uses wrong name_prefix + extension for libraries by default

### ISSUE THREAD

**Issue by madebr** [mesonbuild/meson#7378]
**Describe the bug**
When building libraries using MSVC and the ninja backend, a `lib` `name_prefix` is added.
By default, the `name_prefix` for MSVC should be `""` (the empty string).

Also, the static libraries have the `.a` extension, where it should be `.lib`.

**To Reproduce**

**meson.build**
```
project('example',
    ['c'],
    default_options : ['default_library=static']
)

library('mylibrary',
    ['mysource.c']
)
```
**mysource.c**
```
int my_function(void) {
    return 1337;
}
```
**Expected behavior**
A `mylibrary.lib` file is built.

**Actual behavior**
A `libmylibrary.a` file is built.

**system parameters**
* cross build? no
* os: Windows 10
* compiler: Visual Studio 2017
* python version? 3.8.3
* meson version? 0.54.2
* ninja version? 1.9.0

**Comment by dcbaker** [comment]
Using .a is an explicit design choice. There is no real requirement on what a static library extension needs to be on windows, and we picked .a to differentiate from the interface libraries needed to link against .dlls (which IIRC must be .lib), which is a requirement to be able to build and install both a static and shared library at the same time.

**Comment by dcbaker** [comment]
I can't remember why (or if we should) use `lib` as a prefix, that does seem off to me.

**Comment by madebr** [comment]
It's a weird default for the suffix when using MSVC. It's not expected.
This means that the filename depends on the backend, because the msvc* backends do not do this.
Also, a lot of tools assume `.lib` extensions (e.g. conan), when looking for libraries.

**Comment by TheQwertiest** [comment]
> Using .a is an explicit design choice. There is no real requirement on what a static library extension needs to be on windows

And it is a wrong and a bad design choice. There are no *hard* requirements, no. But! As it was mentioned above not a single IDE or a build-system expects static libraries to have a `*.a` extension. Even meson itself uses `name*.lib` patter to search for static libraries! This behavior violates long established practices on Windows.

I understand, that the ship have sailed and the default behaviour can't be (easily) changed. But, IMO, it should be made possible for the user to make static libraries to generate expected output with a single cmd switch and without having to rename libs manually in the install script.

Generating both libraries is not really common on Windows, but it's usually solved by adding various (alas, horrible) suffixes to the produced output, e.g. `boost-system-vc140-s-gd.lib` vs `boost-system-vc140-gd.lib`.

In other words priorities were mixed up when this design was made: it made common scenario harder to use, and simplified(?) an uncommon scenario.

**Comment by TheQwertiest** [comment]
@dcbaker , `backend:ninja` tag is not needed here, since this problem is present on all backends.

**Comment by TheQwertiest** [comment]
And it's often not possible to use `both_libraries` on windows anyway, because of the different defines that are required for shared and static libraries (most prominent example is `__declspec(export/import)`): https://github.com/mesonbuild/meson/issues/3304#issuecomment-407704983 

Which makes this naming scheme reason kinda moot: "this naming scheme is required for `both_libraries` to work, which actually doesn't work properly in a lot of common scenarios on Windows anyway"

**Comment by SpaceIm** [comment]
What is the status of this issue? It's quite tedious to manually rename static libs after installation.

**Comment by eli-schwartz** [comment]
> Even meson itself uses `name*.lib` patter to search for static libraries!

Sure, Meson searches for .lib, and also for non-MSVC compilers, i.e. mingw GCC, it also searches for dll.a. This goes together, not only with the fact that Meson creates .dll.a, but the fact that the ecosystem also creates .dll.a.

There's a relevant comment in the implementation that describes why we do this. 

> I can't remember why (or if we should) use `lib` as a prefix, that does seem off to me.

@dcbaker, this is also answered by the comment, I think.

Anyway:
https://github.com/mesonbuild/meson/blob/d141b85462fd8560eab9d1d8ae23978426f06907/mesonbuild/build.py#L1976-L1982

This seems pretty relevant. Not only do we use mingw's upstream-preferred file extension when using mingw toolchains, but we also use mingw's preferred file prefix when using mingw toolchains.

> This behavior violates long established practices on Windows.

I love this thought of Windows *having* long established practices. :)

Since MSVC apparently does require you to pass the filename anyway, and doesn't care what you call it, while the extension we do use is the one that mingw expects and insists on... it's not clear to me that this is something which should change.

At least, using "lib" has to go together with using ".a", or we sharply regress for mingw.

Could we use foo.lib instead of libfoo.a? The logic here seems to be that we can, as long as people don't build both shared and static. But if they do, then what? Should we error out saying that "no, you cannot do that, lots of projects don't support both at the same time so we will simply refuse altogether to allow you to even try it"?

**Comment by eli-schwartz** [comment]
If cmake and conan don't look for the filenames that are traditionally used on mingw regardless of build system, that seems like something that might be worth fixing independent of any resolution here.

**Comment by JamBalaya56562** [comment]
To note the current state of this on master (heading toward 1.11):

Since **Meson 1.10.0** there is a built-in `namingscheme` option that directly addresses the single-switch request raised in this thread. Setting `-Dnamingscheme=platform` makes Meson emit platform-native library names on Windows, which for the reporter's exact scenario (`default_library=static` + MSVC) produces `mylibrary.lib` instead of `libmylibrary.a` — no more `lib` prefix and no more `.a` suffix.

The mapping used for the platform scheme:

```python
# mesonbuild/build.py
DEFAULT_STATIC_LIBRARY_NAMES = {
    'unix':    ('lib', 'a'),
    'windows': ('', 'lib'),      # -> mylibrary.lib
    'darwin':  ('lib', 'a'),
    'cygwin':  ('lib', 'a'),
}

DEFAULT_SHARED_LIBRARY_NAMES = {
    'unix':    ('lib', 'so', ''),
    'windows': ('', 'dll', 'dll.lib'),   # -> mylibrary.dll + import mylibrary.dll.lib
    'darwin':  ('lib', 'dylib', ''),
    'cygwin':  ('cyg', 'dll', 'dll.a'),
}
```

So on Windows with `platform`: static libraries become `foo.lib`, shared libraries `foo.dll`, and import libraries `foo.dll.lib` — the `.lib`/`.dll.lib` split keeps the two from colliding, which was the original reason for the `libfoo.a` default (mirrors what Rust does).

**What is resolved:** the recurring ask in this thread — "make it possible to get `foo.lib` on MSVC with a single command-line switch, without renaming in the install script" — is now satisfied via `-Dnamingscheme=platform`. It fixes both the prefix (`lib`) and the extension (`.a`) at once.

**What is NOT changed / is by design:**
- The **default is still `classic`** (`libfoo.a` on MSVC). Out-of-the-box behavior is unchanged, and that default remains intentional for the collision/mingw-compat reasons documented in the FAQ and in the code comment at `build.py:2396-2405`.

  ```python
  # mesonbuild/options.py
  UserComboOption('namingscheme', 'How target file names are formed',
                  'classic', choices=['platform', 'classic']),
  ```

- The 1.10.0 release note states the default is *expected* to switch to `platform` in a future major version, and explicitly **reserves the right to alter how `platform` names its output files** until then — so `platform` is usable today but not yet frozen.
- The FAQ entry ("Why does building my project with MSVC output static libraries called `libfoo.a`?") still only documents the older per-target `name_prefix: ''` / `name_suffix: 'lib'` workaround and does not yet mention `namingscheme=platform`; updating it would probably help future readers who land here.

I'll leave the disposition to the maintainers — flagging this mainly so anyone finding this issue knows there is now a supported, project-wide switch for the requested behavior, while the default naming remains deliberately unchanged.

## Verification
- Release note: `docs/markdown/Release-notes-for-1.10.0.md` -> "Added new `namingscheme` option" (https://mesonbuild.com/Release-notes-for-1.10.0.html)
- Introducing commit: `e500294480a4e1d3e2f236c83d21b279877df5cf` "Add option to specify target naming scheme." (jpakkane, 2025-09-09), with shared-lib support in `8d490ab081dc15abeaece45bacf2ae1bb73e67fd` and docs in `9c1e868b3c588a9c7269babe9381d824cb77e6b5`. Commit dates (Sep 2025) are consistent with the 1.10.0 release notes.
- Code paths verified on master: option definition `mesonbuild/options.py:694`; static naming `mesonbuild/build.py:2387-2425`; shared naming `mesonbuild/build.py:2595-2609`; scheme tables `build.py:165-177`; platform-name selector `build.py:1964-1973`.

**Comment by madebr** [comment]
That should work.

Naming-things-wise, I think `windows` is not entirely accurate. MinGW toolchains use `lib` prefixes, `.a` static suffix and `.dll.a` import library suffix. I'd suggest adding a `mingw` option to address this.

Can this naming option be added to cross files?

**Comment by bonzini** [comment]
> Can this naming option be added to cross files?

Yes it's like any other built-in option.

**Comment by dcbaker** [comment]
We probably should have a mingw platform or mingw subsystem, and then `platform` could just do the right thing for mingw. There's a bunch of cases where I've seen build systems using fragile detection mechanisms like `host_machine.system() == 'windows' and meson.get_compiler('c', native : false).id == 'gcc'`
